pkg/controller/certificaterequests/venafi/venafi.go

```diff
@@ -104,31 +112,51 @@ func (v *Venafi) Sign(ctx context.Context, cr *cmapi.CertificateRequest, issuerO
 		}
 	}
 
-	certPem, err := client.Sign(cr.Spec.CSRPEM, duration, customFields)
+	pickupID, exists := cr.ObjectMeta.Annotations[VenafiPickupIDAnnotation]
 
-	// Check some known error types
-	if err != nil {
-		switch err.(type) {
+	// check if the pickup ID annotation is there, if not set it up.
+	if !exists || pickupID == "" {
+		pickupID, err = client.RequestCertificate(cr.Spec.CSRPEM, duration, customFields)
+		// Check some known error types
+		if err != nil {
+			switch err.(type) {
 
-		case venafiinternal.ErrCustomFieldsType:
-			v.reporter.Failed(cr, err, "CustomFieldsError", err.Error())
-			log.Error(err, err.Error())
+			case venaficlient.ErrCustomFieldsType:
+				v.reporter.Failed(cr, err, "CustomFieldsError", err.Error())
+				log.Error(err, err.Error())
 
-			return nil, nil
+				return nil, nil
 
-		case endpoint.ErrCertificatePending:
-			message := "Venafi certificate still in a pending state, the request will be retried"
+			default:
+				message := "Failed to request venafi certificate"
 
-			v.reporter.Pending(cr, err, "IssuancePending", message)
-			log.Error(err, message)
+				v.reporter.Failed(cr, err, "RequestError", message)
+				log.Error(err, message)
+
+				return nil, err
+			}
+		}
+
+		v.reporter.Pending(cr, err, "IssuancePending", "Venafi certificate is requested")
+
+		metav1.SetMetaDataAnnotation(&cr.ObjectMeta, VenafiPickupIDAnnotation, pickupID)
+		_, err = v.cmClient.CertmanagerV1alpha2().CertificateRequests(cr.GetNamespace()).Update(ctx, cr, metav1.UpdateOptions{})
```

Comment:
Calling Update here will cause issues when the 'reporter' also tries to update the CertificateRequest as the ResourceVersion will have changed. We need to avoid ever calling Update on the resource being synced within these Sync functions. Again, collapsing all this code into an actual controller will make it _far_ easier to notice and and work around this sort of thing.

It's especially complex here as we need to call *Update* and not *UpdateStatus*...

We definitely need to come up with a solution to the resourceVersion conflict this will cause 👀

Reply:
Also: if this Update call fails, how will the controller behave/will it recover? Is there any other way to 'recover' an in progress request?

Reply:
we could add annotation changes in https://github.com/jetstack/cert-manager/blob/master/pkg/controller/certificaterequests/sync.go#L160:1